{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d3e6383d-6bb1-4185-a99e-f1eb0f5adc2a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Tutorial\n",
    "\n",
    "<a href=\"?print-pdf\">print view</a>  \n",
    "<a href=\"tutorial.ipynb\" download>notebook</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa3cb82-0eb0-4eb5-8190-1980b3e2e70b",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Experiment: Arabidopsis thaliana (Thale Cress)\n",
    "- Comparing wild type (WT) and atrx-1 mutant plants to analyze how the loss of function of ATRX protein results in changes in gene expression.\n",
    "- The ATRX protein is a histone chaperone known to be an important player in the regulation of gene expression.\n",
    "- RNA was isolated from three WT replicates and three mutant replicates.\n",
    "- RNA-seq was carried out on an Illumina Hiseq 2500.\n",
    "- The sequencing reads are paired-end data, hence we have 2 files per replicate.\n",
    "- \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c238a0a5-ebfa-4e51-8126-52f41103162a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Create directory in which we will be doing the analysis \n",
    "- **create directory: mkdir -p tutorial/RNAseq**\n",
    "- **Go to the directory: cd tutorial/RNAseq**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b12b62-187a-490c-bfd5-85999240aceb",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Download the data:\n",
    "- **wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR442/003/SRR4420293/SRR4420293_1.fastq.gz**\n",
    "- **wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR442/003/SRR4420293/SRR4420293_2.fastq.gz**\n",
    "- **wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR442/006/SRR4420296/SRR4420296_1.fastq.gz**\n",
    "- **wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR442/006/SRR4420296/SRR4420296_2.fastq.gz**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a9e69e0-7dde-4a95-b63f-2c4e9064f246",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Get the Genome reference file and associated GTF/GFF file for Arabidopsis. \n",
    "- **Create a directory for the genome: mkdir Genome**\n",
    "- **Go to the directory: cd Genome**\n",
    "- **wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_genomic.fna.gz**\n",
    "- **wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_genomic.gtf.gz**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6ff3277-c1f5-404d-a7ce-bdc54f92da82",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Unzip the files \n",
    "- **gunzip GCF_000001735.4_TAIR10.1_genomic.fna.gz**\n",
    "- **gunzip GCF_000001735.4_TAIR10.1_genomic.gtf.gz**\n",
    "- **Check if they have been unzipped**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00bb33f6-0bae-48f9-9650-a96bcf809669",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Quality Check\n",
    "- **Create a folder for fastqc files: mkdir fastqcReports**\n",
    "- **Check the number of cores:**\n",
    "    - **import os**\n",
    "    - **os.cpu_count()**\n",
    "- **fastqc -t 4 -o ./fastqcReports/ ./*.gz**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4757a517-7114-4f94-9ef2-4bc79b4f428c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Detour: When more than one QC file: Compile all the results together using multiqc \n",
    "- **install multiqc: pip3 install multiqc**\n",
    "- **multiqc -p -o ./fastqcReports/ ./fastqcReports/ --title \"Arabidopsis_RNAseq\"**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22f50ebf-b23b-437d-9622-1112e460e17a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Detour: adapter trimming \n",
    "- **installation: conda install bioconda::trimmomatic**\n",
    "- **mkdir adapter**\n",
    "- **cd adapter**\n",
    "- **download the adapter fasta:**\n",
    "    - **wget https://github.com/usadellab/Trimmomatic/blob/main/adapters/TruSeq3-PE.fa**\n",
    "- **apply trimmomatic**\n",
    "- <code>trimmomatic PE -threads 4 \\\n",
    "         SRR4420293_1.fastq.gz SRR4420293_2.fastq.gz \\\n",
    "         SRR4420293_1_trimmed_paired.fastq.gz SRR4420293_1_trimmed_unpaired.fastq.gz \\\n",
    "         SRR4420293_2_trimmed_paired.fastq.gz SRR4420293_2_trimmed_unpaired.fastq.gz \\\n",
    "         ILLUMINACLIP:./adapter/TruSeq3-PE.fa:2:30:10 \\\n",
    "         LEADING:3 TRAILING:3 \\\n",
    "         SLIDINGWINDOW:4:15 \\\n",
    "         MINLEN:36</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "698a3949-ba67-4e0b-aa2a-2f09af915ca4",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Aligning the reads:\n",
    "- **Install hisat2: conda install -c bioconda hisat2**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95d1ab2a-f1d1-46d5-af7b-c76054950546",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Extract splice sites and exons from GTF (optional but recommended)\n",
    "- **hisat2_extract_splice_sites.py ./Genome/GCF_000001735.4_TAIR10.1_genomic.gtf > ./splicesites.txt**\n",
    "- **hisat2_extract_exons.py ./Genome/GCF_000001735.4_TAIR10.1_genomic.gtf > ./exons.txt**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c18da63e-b9fc-4f1f-9f1c-e9d276e7e9ec",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Build HISAT2 index with splice site information\n",
    "\n",
    "<code>hisat2-build \\\\\n",
    "       -p 4 \\\\\n",
    "       --ss ./splicesites.txt \\\\\n",
    "      --exon ./exons.txt \\\\\n",
    "      ./GCF_000001735.4_TAIR10.1_genomic.fna \\\\\n",
    "      hisat_index/TAIR10</code>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9af256e4-cec6-4077-a073-677c258aa41e",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Now run HISAT2 with the index PREFIX to align the samples\n",
    "<code>hisat2 -p 4 \\\\\n",
    "    -x ./genome/hisat_index/TAIR10 \\\\\n",
    "    -1 SRR4420293_1.fastq.gz \\\\\n",
    "    -2 SRR4420293_2.fastq.gz \\\\\n",
    "    -S aligned_reads/SRR4420293.sam \\\\\n",
    "    --summary-file ./aligned_reads/SRR4420293_alignment_summary.txt </code>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a479561-7f78-4140-b6f7-528035ecabb2",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# install samtools\n",
    "- **conda install -c bioconda samtools**\n",
    "- **verify the installation: samtools --version**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e12be82-677e-4bab-810d-5660451f6b58",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Convert to sorted BAM\n",
    "<code>samtools view -@ 4 -bS aligned_reads/SRR4420293.sam | \\\\\n",
    "samtools sort -@ 4 -o aligned_reads/SRR4420293.sorted.bam && \\\\\n",
    "samtools index aligned_reads/SRR4420293.sorted.bam && \\\\\n",
    "rm aligned_reads/SRR4420293.sam</code>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1b461206-d703-4f8c-9fab-f333d0e8c3de",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# install featurecounts\n",
    "- **conda install -c bioconda subread**\n",
    "- **verify installation: featureCounts -v**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b51d6d8d-d393-438d-9d37-8be2c4d23e1a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Let's count the features \n",
    "\n",
    "<code>featureCounts \\\\\n",
    "    -a ./genome/GCF_000001735.4_TAIR10.1_genomic.gtf \\\\\n",
    "    -o ./gene_counts/SRR4420293_counts.txt \\\\\n",
    "    -p \\\\\n",
    "    -t exon \\\\\n",
    "    -g gene_id \\\\\n",
    "    -T 4 \\\\\n",
    "    ./aligned_reads/SRR4420293.sorted.bam</code>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67410e6c-aa28-4bb3-9e27-7a98657d4c8c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "source": [
    "# Quiz\n",
    "- **Convert and save the counts txt file(s) into a csv file**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
